modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageRaftGroupEventsListener.java

```diff
@@ -0,0 +1,247 @@
+import org.apache.ignite.network.TopologyEventHandler;
+import org.apache.ignite.network.TopologyService;
+
+/**
+ * Raft Group Events listener that registers Logical Topology listeners for removing learners that have left the topology.
+ */
+public class MetaStorageRaftGroupEventsListener implements RaftGroupEventsListener {
+    private static final IgniteLogger LOG = Loggers.forClass(MetaStorageManagerImpl.class);
+
+    private final IgniteSpinBusyLock busyLock;
+
+    private final TopologyService topologyService;
+
+    private final LogicalTopology logicalTopology;
+
+    private final CompletableFuture<MetaStorageServiceImpl> metaStorageSvcFut;
+
+    /**
+     * Future required to enable serialized processing of topology events.
+     *
+     * <p>While Logical Topology events are linearized, we also concurrently react to Physical Topology changes and/or update Raft
+     * configuration after a new leader has been elected.
+     *
+     * <p>Multi-threaded access is guarded by {@code serializationFutureMux}.
+     */
+    private CompletableFuture<Void> serializationFuture = null;
+
+    private final Object serializationFutureMux = new Object();
+
+    /**
+     * The most recent version of the Logical Topology that this listener has observed.
+     */
+    private volatile long lastProcessedLogicalTopologyVersion = -1;
+
+    MetaStorageRaftGroupEventsListener(
+            IgniteSpinBusyLock busyLock,
+            ClusterService clusterService,
+            LogicalTopology logicalTopology,
+            CompletableFuture<MetaStorageServiceImpl> metaStorageSvcFut
+    ) {
+        this.busyLock = busyLock;
+        this.topologyService = clusterService.topologyService();
+        this.logicalTopology = logicalTopology;
+        this.metaStorageSvcFut = metaStorageSvcFut;
+    }
+
+    @Override
+    public void onLeaderElected(long term) {
+        registerTopologyEventListeners();
+
+        // Update learner configuration in case we missed some topology updates.
+        executeIfLeader(service -> removeLearners(service.raftGroupService(), logicalTopology.getLogicalTopology()));
+    }
+
+    private void registerTopologyEventListeners() {
+        // Try to remove a Meta Storage Learner if a node leaves the Physical Topology. This handles the case when a node joins, but leaves
+        // before sending the Join Ready request (i.e. before it gets added to the Logical Topology).
+        topologyService.addEventHandler(new TopologyEventHandler() {
+            @Override
+            public void onDisappeared(ClusterNode member) {
+                executeIfLeader(service -> removeLearners(service.raftGroupService(), logicalTopology.getLogicalTopology()));
+            }
+        });
+
+        logicalTopology.addEventListener(new LogicalTopologyEventListener() {
+            // Try to add a node as a Learner when it appears in the Logical Topology. Even though usually remote nodes add themselves as
```

Comment:
Could you please elaborate the flow in a greater detail? I didn't get the case.

Reply:
Normal flow:
1. A node gets discovered by the CMG leader using the Physical Topology.
2. The node gets validated on the CMG leader.
3. The CMG leader sends MS node IDs to the joining node.
4. The joining node starts a local Raft server and executes `addLearners` to add itself as a learner.
5. The joining node executes `onJoinReady` and gets added to the Logical Topology.

Now imagine that after that a network partition happens for a long enough time for the node to get kicked out of the Logical Topology. As soon as that happens, it gets removed from the MS learners as well. When the network partition is fixed, the node will be added back to the Logical Topology, but it will not be added as MS learner (because the node was not restarted and the code that starts the Raft server will not be executed).

Reply:
Got it. I would expect that joining process should be aborted if joining node was removed from physical topology and restarted from the very beginning if needed. Otherwise we may have inconsistencies in node id, etc.

Reply:
Do you suggest that if a node leaves the logical topology for whatever reason, should re-run the joining process even if it was not restarted?